From cdfdcb4d120a2c898684bc4e972b01dd462ed5a1 Mon Sep 17 00:00:00 2001
From: Steve Lhomme <robux4@ycbcr.xyz>
Date: Mon, 5 Sep 2022 13:34:41 +0200
Subject: [PATCH] Export the proper C++ runtime library

With gcc it's libstdc++, with clang it's libc++. So let CMake give it to use
and use all the libraries it needs, except some noise from clang toolchains.
---
 gme/CMakeLists.txt | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/gme/CMakeLists.txt b/gme/CMakeLists.txt
index fc2db68..301f8c6 100644
--- a/gme/CMakeLists.txt
+++ b/gme/CMakeLists.txt
@@ -239,8 +239,29 @@ endforeach()
 
 add_library(gme_deps INTERFACE)
 
-## FIXME: Properly find the C++ library !!!
-set(PC_LIBS -lstdc++)
+foreach(LIB ${CMAKE_CXX_IMPLICIT_LINK_LIBRARIES})
+    if(IS_ABSOLUTE ${LIB} AND EXISTS ${LIB})
+        list(APPEND IMPLICITS_LIST "${LIB}")
+    elseif(LIB MATCHES "-l:lib.*.a")
+        string(LENGTH ${LIB} LIBLEN)
+        math(EXPR LIBLEN "${LIBLEN}-8")
+        string(SUBSTRING ${LIB} 6 ${LIBLEN} DIRECT_LIB)
+        list(APPEND IMPLICITS_LIST "-l${DIRECT_LIB}")
+    elseif(LIB MATCHES "-l.*")
+        list(APPEND IMPLICITS_LIST "${LIB}")
+    else()
+        list(APPEND IMPLICITS_LIST "-l${LIB}")
+    endif()
+endforeach()
+if(IMPLICITS_LIST)
+    # blacklist of libraries that should not be in Libs.private
+    list(REMOVE_ITEM IMPLICITS_LIST "-lmingwex"
+         "-lmingw32" "-lmoldname" "-lmsvcrt" "-ladvapi32" "-lshell32"
+         "-luser32" "-lkernel32")
+    string(REPLACE ";" " " PC_LIBS "${IMPLICITS_LIST}")
+else()
+    set(PC_LIBS "")
+endif(IMPLICITS_LIST)
 
 if(GME_ZLIB)
     if(ZLIB_FOUND)
-- 
2.50.1 (Apple Git-155)

